store/src/java/com/zimbra/cs/service/mail/WaitSetRequest.java

```diff
@@ -308,30 +310,56 @@ public static void staticHandle(WaitSetReq req, Map<String, Object> context, Wai
                 }
 
                 if(accountMods!= null && accountMods.modified != null) {
+                    //aggregate tag changes so they are sent to each folder we are interested in
+                    List<ModifyTagNotification> tagMods = new ArrayList<ModifyTagNotification>();
+                    for(Object maybeTagChange : accountMods.modified.values()) {
+                        if(maybeTagChange instanceof Change) {
+                            Object maybeTag = ((Change) maybeTagChange).what;
+                            if(maybeTag != null && maybeTag instanceof Tag) {
+                                Tag tag = (Tag) maybeTag;
+                                tagMods.add(new ModifyTagNotification(tag.getIdInMailbox(), tag.getName(), ((Change) maybeTagChange).why));
+                            }
+                        }
+                    }
                     for(Object mod : accountMods.modified.values()) {
                         if(mod instanceof Change) {
                             Object what = ((Change) mod).what;
                             if(what != null && what instanceof BaseItemInfo) {
                                 BaseItemInfo itemInfo = (BaseItemInfo)what;
                                 Integer folderId = itemInfo.getFolderIdInMailbox();
-                                if(folderInterests != null && !folderInterests.contains(folderId)) {
-                                    continue;
+                                if (itemInfo instanceof Folder) {
+                                    Integer itemId = itemInfo.getIdInMailbox();
+                                    if(folderInterests != null &&
+                                            !folderInterests.contains(folderId) &&
+                                            !folderInterests.contains(itemId)) {
+                                        continue;
+                                    }
+                                    if (!tagMods.isEmpty()) {
+                                        PendingFolderModifications folderMods = getFolderMods(itemId, folderMap);
+                                        for (ModifyTagNotification modTag: tagMods) {
+                                            folderMods.addModifiedTag(modTag);
+                                        }
+                                    }
+                                } else if (!(itemInfo instanceof Tag)){
+                                    if(folderInterests != null && !folderInterests.contains(folderId)) {
+                                        continue;
+                                    }
+                                    getFolderMods(folderId, folderMap).addModifiedMsg(getModifiedItemSOAP(itemInfo, ((Change) mod).why));
```

Comment:
I've moved getFolderMods to JaxbUtil, because it is now also used by SoapSession